### PR TITLE
improve gateway light class

### PR DESCRIPTION
**Pull request by starkillerOG**
cleanup the gateway light class and add missing functionality needed for HomeAssistint

- added docstrings to explain all functionality.
- now contains these functions:
  - rgb_status
  - night_light_status
  - set_rgb
  - set_night_light
  - set_rgb_brightness
  - set_night_light_brightness
  - set_rgb_color
  - set_night_light_color
  - set_rgb_using_name
  - set_night_light_using_name

### CODE REVIEW

**Comment by starkillerOG** [comment]
@rytilahti I just fully tested this code and works just fine.
I think this can be merged.

File: miio/gateway.py [hunk 1]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
-        """Unknown."""
-        # Returns 0 when light is off?"""
-        # looks like this is the same as get_rgb
-        # id': 65064, 'method': 'set_night_light_rgb', 'params': [419407616]}
-        # {'method': 'props', 'params':
-        # {'light': 'on', 'from.light': '4,,,'}, 'id': 88457} ?!
-        return self._gateway.send("get_night_light_rgb")
+    def rgb_status(self):
+        """Get current status of the light (always correct)."""
+        # Returns {"is_on": false, "brightness": 0, "rgb": (0, 0, 0)} when light is off
```
Comment (rytilahti) [inline miio/gateway.py:651]: I think this commentary could go directly to the docstring:
```
"""Get current status of the light.

Example:
   {"is_on": false, "brightness": 0, "rgb": (0, 0, 0)}
"""

It is also unclear what does "always correct" mean in this context, so that should be clarified (or removed, if unnecessary).

File: miio/gateway.py [hunk 2]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
+        else:
+            is_on = False
+        return {"is_on": is_on, "brightness": brightness, "rgb": rgb}
+
+    @command()
+    def night_light_status(self):
+        """Get status of the night light."""
+        # The night light controls the same LEDs as the rgb light, but has a seprate memory for brightness and color.
+        # Changing the rgb light does not effect the stored state of the night light, while changing the night light does effect the state of the rgb light.
+        # Therefore this command only gives the correct status of the LEDs if the last command was a night light command and not a rgb light command, otherwise it gives the stored values of the night light.
```
Comment (rytilahti) [inline miio/gateway.py:666]: Include this description also in the docstring, that way it is visible in the apidocs.

File: miio/gateway.py [hunk 3]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
+    def rgb_status(self):
+        """Get current status of the light (always correct)."""
+        # Returns {"is_on": false, "brightness": 0, "rgb": (0, 0, 0)} when light is off
+        state_int = self._gateway.send("get_rgb").pop()
+        brightness = int_to_brightness(state_int)
+        rgb = int_to_rgb(state_int)
+        if brightness > 0:
+            is_on = True
+        else:
+            is_on = False
```
Comment (rytilahti) [inline miio/gateway.py:658]: This can be simplified by using `is_on = brightness > 0`. Same also below for the night light.

File: miio/gateway.py [hunk 4]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
+        # Therefore this command only gives the correct status of the LEDs if the last command was a night light command and not a rgb light command, otherwise it gives the stored values of the night light.
+        # Returns {"is_on": false, "brightness": 0, "rgb": (0, 0, 0)} when night light is off
+        state_int = self._gateway.send("get_night_light_rgb").pop()
+        brightness = int_to_brightness(state_int)
+        rgb = int_to_rgb(state_int)
+        if brightness > 0:
+            is_on = True
+        else:
+            is_on = False
+        return {"is_on": is_on, "brightness": brightness, "rgb": rgb}
```
Comment (rytilahti) [inline miio/gateway.py:675]: Add a newline before the return to make the code more readable.

File: miio/gateway.py [hunk 5]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
+
+    @command(
+        click.argument("brightness", type=int),
+        click.argument("rgb", type=(int, int, int)),
+    )
+    def set_rgb(self, brightness, rgb):
```
Comment (rytilahti) [inline miio/gateway.py:681]:
```suggestion
    def set_rgb(self, brightness: int, rgb: Tuple[int, int, int]):
```

Comment (rytilahti) [reply miio/gateway.py:681]: Type annotations should also be added to other other functions.

File: miio/gateway.py [hunk 6]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
+        """Set night light brightness (0-100)."""
+        if 100 < brightness < 0:
+            raise Exception("Brightness must be between 0 and 100")
+        current_color = self.night_light_status()["rgb"]
+        return self.set_night_light(brightness, current_color)
 
     @command(click.argument("color_name", type=str))
-    def set_night_light_color(self, color_name):
-        """Set night light color using color name (red, green, etc)."""
+    def set_rgb_color(self, color_name):
```
Comment (rytilahti) [inline miio/gateway.py:712]: rgb consists of three separate components (r, g, and b), so shouldn't this rather be `set_color`?

File: miio/gateway.py [hunk 7]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
+        """Set gateway light color using color name (red, green, etc)."""
```
Comment (rytilahti) [inline miio/gateway.py:713]: Is the list of colors static and are the valid values known? Maybe this should be enforced using an enum (or a list)?

File: miio/gateway.py [hunk 8]
```diff
@@ -646,71 +646,112 @@ class GatewayLight(GatewayDevice):
...
-        )
-        return self._gateway.send("set_night_light_rgb", [brightness_and_color])
+        current_brightness = self.rgb_status()["brightness"]
+        return self.set_rgb(current_brightness, color_map[color_name])
 
     @command(click.argument("color_name", type=str))
-    def set_color(self, color_name):
-        """Set gateway lamp color using color name (red, green, etc)."""
+    def set_night_light_color(self, color_name):
+        """Set night light color using color name (red, green, etc)."""
```
Comment (rytilahti) [inline miio/gateway.py:725]: See my comment above wrt color names.

File: miio/gateway.py (hunk 6 above)
Comment (starkillerOG) [reply miio/gateway.py:712]: I wanted to be consistant between "rgb" and "night_light"
All functions begin with `set_rgb` or `set_night_light`

So I think this is the correct name

File: miio/gateway.py (hunk 1 above)
Comment (starkillerOG) [reply miio/gateway.py:657]: done, you are right!

File: miio/gateway.py (hunk 2 above)
Comment (starkillerOG) [reply miio/gateway.py]: done

File: miio/gateway.py (hunk 3 above)
Comment (starkillerOG) [reply miio/gateway.py]: done, thanks!

File: miio/gateway.py (hunk 4 above)
Comment (starkillerOG) [reply miio/gateway.py:681]: I added newlines to all functions before the retrun's

File: miio/gateway.py (hunk 5 above)
Comment (starkillerOG) [reply miio/gateway.py]: thanks, did it for all functions of this PR

File: miio/gateway.py (hunk 7 above)
Comment (starkillerOG) [reply miio/gateway.py:723]: This is a static dict that is defined on line 26:
https://github.com/rytilahti/python-miio/blob/79e61fbba512a79b569fc5cf5d14a2b5e7730c01/miio/gateway.py#L26

I actually did not program this and it was already in the code like this.
This will not be used for HomeAssistant since there you just set RGB values and HomeAssistant has its own color names defined.
I don't think these functions will be used much.
basically you can now specify "red", "green", "blue", "white", "yellow", "orange", "aqua", "olive" or "purple"
I will put it in the docstrings

**Comment by starkillerOG** [comment]
@rytilahti any more suggestions?

File: miio/gateway.py (hunk 7 above)
Comment (rytilahti) [reply miio/gateway.py]: Ah, okay. Do not add them to the docstring, but add a comment saying that `color_map` variable in the source holds the valid values. This avoids repeating the list :-)

File: miio/gateway.py (hunk 6 above)
Comment (rytilahti) [reply miio/gateway.py]: Fair enough!

File: miio/gateway.py [hunk 9]
```diff
@@ -647,69 +647,79 @@ class GatewayLight(GatewayDevice):
...
-        # The night light controls the same LEDs as the rgb light, but has a seprate memory for brightness and color.
-        # Changing the rgb light does not effect the stored state of the night light, while changing the night light does effect the state of the rgb light.
-        # Therefore this command only gives the correct status of the LEDs if the last command was a night light command and not a rgb light command, otherwise it gives the stored values of the night light.
-        # Returns {"is_on": false, "brightness": 0, "rgb": (0, 0, 0)} when night light is off
+        """
+        Get status of the night light.
+
+        The night light controls the same LEDs as the rgb light, but has a seprate memory for brightness and color.
+        Changing the rgb light does not effect the stored state of the night light, while changing the night light does effect the state of the rgb light.
+        Therefore this command only gives the correct status of the LEDs if the last command was a night light command and not a rgb light command, otherwise it gives the stored values of the night light.
```
Comment (rytilahti) [inline miio/gateway.py:672]: Some sort of explanation to the class' docstring would be great, I think this could (at least partially) belong there as it description itself is not just about the night light but the whole light support.

What do you think?

File: miio/gateway.py (hunk 7 above)
Comment (starkillerOG) [reply miio/gateway.py]: Alright, I changed it

File: miio/gateway.py (hunk 9 above)
Comment (starkillerOG) [reply miio/gateway.py:672]: You are right!
I will change it, just a moment

Comment (starkillerOG) [reply miio/gateway.py]: I made a more clear explanation in the class string, and removed the duplicate info here.

**Comment by starkillerOG** [comment]
@rytilahti are the docstrings more clear now?

**Comment by starkillerOG** [comment]
@rytilahti all test pass again.
Anything else I can improve on this?

File: miio/gateway.py [hunk 10]
```diff
@@ -643,74 +643,132 @@ def set_default_music(self):
 
 
 class GatewayLight(GatewayDevice):
-    """Light controls for the gateway."""
+    """
+    Light controls for the gateway.
+
+    The gateway LEDs can be controlled using 'rgb' or 'night_light' methods.
+    The 'night_light' methods control the same LEDs as the 'rgb' methods, but has a seprate memory for brightness and color.
```
Comment (rytilahti) [inline miio/gateway.py:650]:
```suggestion
    The 'night_light' methods control the same light as the 'rgb' methods, but has a separate memory for brightness and color.
```

File: miio/gateway.py [hunk 11]
```diff
@@ -643,74 +643,132 @@ def set_default_music(self):
...
+    Changing the 'rgb' light does not effect the stored state of the 'night_light', while changing the 'night_light' does effect the state of the 'rgb' light.
```
Comment (rytilahti) [inline miio/gateway.py:651]:
```suggestion
    Changing the 'rgb' light does not affect the stored state of the 'night_light', while changing the 'night_light' does effect the state of the 'rgb' light.
```

File: miio/gateway.py [hunk 12]
```diff
@@ -643,74 +643,132 @@ def set_default_music(self):
...
-        # Returns 0 when light is off?"""
-        # looks like this is the same as get_rgb
-        # id': 65064, 'method': 'set_night_light_rgb', 'params': [419407616]}
-        # {'method': 'props', 'params':
-        # {'light': 'on', 'from.light': '4,,,'}, 'id': 88457} ?!
-        return self._gateway.send("get_night_light_rgb")
+    def rgb_status(self):
+        """
+        Get current status of the light.
+        Always represents the current status of the LEDs as oposed to 'night_light_status'.
```
Comment (rytilahti) [inline miio/gateway.py:658]:
```suggestion
        Always represents the current status of the light as opposed to 'night_light_status'.
```

File: miio/gateway.py [hunk 13]
```diff
@@ -643,74 +643,132 @@ def set_default_music(self):
...
+        rgb = int_to_rgb(state_int)
+        is_on = brightness > 0
+
+        return {"is_on": is_on, "brightness": brightness, "rgb": rgb}
+
+    @command()
+    def night_light_status(self):
+        """
+        Get status of the night light.
+        This command only gives the correct status of the LEDs if the last command was a 'night_light' command and not a 'rgb' light command, otherwise it gives the stored values of the 'night_light'.
```
Comment (rytilahti) [inline miio/gateway.py:675]:
```suggestion
        This command only gives the correct status of the light if the last command was a 'night_light' command and not a 'rgb' light command, otherwise it gives the stored values of the 'night_light'.
```

Am I interpreting this correctly, that if any rgb light commands will overwrite the night light state?

File: miio/gateway.py [hunk 14]
```diff
@@ -643,74 +643,132 @@ def set_default_music(self):
...
+        rgb = int_to_rgb(state_int)
+        is_on = brightness > 0
+
+        return {"is_on": is_on, "brightness": brightness, "rgb": rgb}
+
+    @command(
+        click.argument("brightness", type=int),
+        click.argument("rgb", type=(int, int, int)),
+    )
+    def set_rgb(self, brightness: int, rgb: (int, int, int)):
```
Comment (rytilahti) [inline miio/gateway.py:691]:
```suggestion
    def set_rgb(self, brightness: int, rgb: Tuple[int, int, int]):
```

Also, you need to import `Tuple` from the typing module for this.

File: miio/gateway.py [hunk 15]
```diff
@@ -643,74 +643,132 @@ def set_default_music(self):
...
+        """Set gateway light using brightness and rgb tuple."""
+        brightness_and_color = brightness_and_color_to_int(brightness, rgb)
+
+        return self._gateway.send("set_rgb", [brightness_and_color])
+
+    @command(
+        click.argument("brightness", type=int),
+        click.argument("rgb", type=(int, int, int)),
```
Comment (rytilahti) [inline miio/gateway.py:699]: I'm not sure if such a type is accepted by click, but I think this is acceptable for now considering the cli functionality for gateway is not working anyway at the moment..

File: miio/gateway.py [hunk 16]
```diff
@@ -643,74 +643,132 @@ def set_default_music(self):
...
+    )
+    def set_night_light(self, brightness: int, rgb: (int, int, int)):
```
Comment (rytilahti) [inline miio/gateway.py:701]: See my previous comment about `Tuple`.

File: miio/gateway.py (hunk 13 above)
Comment (starkillerOG) [reply miio/gateway.py:675]: The other way around.
issuing ether a rgb light command or night light command will always change the LEDs to the values in the command.
the rgb_status will then be updated to the actual value of the LEDs

However if a rgb light command is sent, the night_light_status does not change (only the rgb_status changes)
If a night light command is sent, both the nigth_light_status and the rgb_status will change

File: miio/gateway.py (hunk 15 above)
Comment (starkillerOG) [reply miio/gateway.py:699]: Not sure about cli, I actually do not use that, so I try my best to keep everything in style with the rest, but I am not sure about how the cli works with things like this.

**Comment by starkillerOG** [comment]
@rytilahti done, ready to be merged
